internal/report/snapshot/snapshot_buffers.go

```diff
@@ -0,0 +1,255 @@
+// Copyright  observIQ, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snapshot
+
+import (
+	"fmt"
+	"sync"
+
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+// LogBuffer is a buffer for plog.Logs
+type LogBuffer struct {
+	mutex     sync.Mutex
+	buffer    []plog.Logs
+	idealSize int
+}
+
+// NewLogBuffer creates a logBuffer with the ideal size set
+func NewLogBuffer(idealSize int) *LogBuffer {
+	return &LogBuffer{
+		buffer:    make([]plog.Logs, 0),
+		idealSize: idealSize,
+	}
+}
+
+// Len counts the number of log records in all Log payloads in buffer
+func (l *LogBuffer) Len() int {
+	size := 0
+	for _, ld := range l.buffer {
+		size += ld.LogRecordCount()
+	}
+
+	return size
+}
+
+// Add adds the new log payload and adjust buffer to keep ideal size
+func (l *LogBuffer) Add(ld plog.Logs) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	logSize := ld.LogRecordCount()
+	bufferSize := l.Len()
+	switch {
+	// The number of logs is more than idealSize so reset this to just this log set
+	case logSize > l.idealSize:
+		l.buffer = []plog.Logs{ld}
+
+	// Haven't reached idealSize yet so add this
+	case logSize+bufferSize < l.idealSize:
+		l.buffer = append(l.buffer, ld)
+
+	// Adding this will put us over idealSize so and add the new logs.
+	// Only remove the oldest if it does not bring buffer under idealSize
+	case logSize+bufferSize >= l.idealSize:
+		l.buffer = append(l.buffer, ld)
+
+		// If removing the oldest item does not put us under the ideal size then it's ok to do so
+		oldest := l.buffer[0]
+		newBufferSize := logSize + bufferSize
+		if newBufferSize-oldest.LogRecordCount() >= l.idealSize {
+			l.buffer = l.buffer[1:]
+		}
+	}
```

Comment:
It looks like there might be some potential for this to inflate quite a bit over the ideal size here.

e.g. You have a 100 `plog.Logs` with 1 record each, then we start getting `plog.Logs` with 100 records each, we could potentially be holding 100 * 100 = 10000 individual records, when ideal size is only 100. This might not be too uncommon in a bursty environment.

I think the right thing to do here is loop, removing `plog.Logs` until you have only one `plog.Logs` (the one you are adding), or you are under the ideal size.

I have similar concerns with the other buffers, since it's all essentially the same logic.

Reply:
So the logic here is I want to have `idealSize` or larger. Once I reach a buffer of `idealSize` I never want it to be less than that. Since I had 100 logs at one point I should never have less than that. The logic is a bit rough but this makes sure we can always send at least `idealSize` number of items to BPOP if we have them. If we have way more BPOP will filter accordingly. 100 is a fairly small number so I think at larger throughput rates the first case where we just replace the entire buffer with the new payload will be the most common case.

Reply:
i still think what brandon is saying has merit. this logic could be improved in the future by possibly removing more than a single entry from the list. a comment stating that the logic could be improved upon at least would be nice. unless i'm missing the point and you don't think it could be improved all that much.